presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperatorBlockUtil.java

```diff
@@ -13,32 +13,304 @@
 
 final class UnnestOperatorBlockUtil
 {
     private UnnestOperatorBlockUtil() {}
 
-    private static final int DEFAULT_CAPACITY = 64;
-    // See java.util.ArrayList for an explanation
-    static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+    static Block insertNull(Block block, Type type)
+    {
+        if (block instanceof DictionaryBlock) {
+            return insertNullToDictionaryBlock((DictionaryBlock) block, type);
+        }
+
+        if (block instanceof RunLengthEncodedBlock) {
+            return insertNullToRleBlock((RunLengthEncodedBlock) block, type);
+        }
+
+        if (block instanceof ByteArrayBlock) {
+            return insertNullToByteArrayBlock((ByteArrayBlock) block);
+        }
+
+        if (block instanceof ShortArrayBlock) {
+            return insertNullToShortArrayBlock((ShortArrayBlock) block);
+        }
+
+        if (block instanceof IntArrayBlock) {
+            return insertNullToIntArrayBlock((IntArrayBlock) block);
+        }
+
+        if (block instanceof LongArrayBlock) {
+            return insertNullToLongArrayBlock((LongArrayBlock) block);
+        }
+
+        if (block instanceof Int128ArrayBlock) {
+            return insertNullToInt128ArrayBlock((Int128ArrayBlock) block);
+        }
+
+        if (block instanceof VariableWidthBlock) {
+            return insertNullToVariableWidthBlock((VariableWidthBlock) block);
+        }
+
+        if (block instanceof ArrayBlock) {
+            return insertNullToArrayBlock((ArrayBlock) block);
+        }
+
+        if (block instanceof MapBlock) {
+            return insertNullToMapBlock((MapBlock) block, (MapType) type);
+        }
+
+        if (block instanceof RowBlock) {
+            return insertNullToRowBlock((RowBlock) block);
+        }
+
+        throw new PrestoException(NOT_SUPPORTED, "Cannot insert null to " + block.getClass());
+    }
 
-    // Copied from io.prestosql.spi.block.BlockUtil#calculateNewArraySize
-    static int calculateNewArraySize(int currentSize)
+    private static Block insertNullToDictionaryBlock(DictionaryBlock block, Type type)
     {
-        // grow array by 50%
-        long newSize = (long) currentSize + (currentSize >> 1);
+        int positionCount = block.getPositionCount();
+        Block dictionary = block.getDictionary();
 
-        // verify new size is within reasonable bounds
-        if (newSize < DEFAULT_CAPACITY) {
-            newSize = DEFAULT_CAPACITY;
+        int[] ids = new int[positionCount + 1];
+        for (int i = 0; i < positionCount; i++) {
+            ids[i] = block.getId(i);  // absolute id
         }
-        else if (newSize > MAX_ARRAY_SIZE) {
-            newSize = MAX_ARRAY_SIZE;
-            if (newSize == currentSize) {
-                throw new IllegalArgumentException(format("Can not grow array beyond '%s'", MAX_ARRAY_SIZE));
+
+        NullElementFinder nullElementFinder = new NullElementFinder();
```

Comment:
Using `NullElementFinder` is not beneficial here. Let's just write out the logic:

```
        int nullIndex = findNull(dictionary);

    private static int findNull(Block block)
    {
        int nullIndex = -1;
        if (block.mayHaveNull()) {
            for (int i = 0; i < block.getPositionCount(); i++) {
                if (block.isNull(i)) {
                    nullIndex = i;
                }
            }
        }
        return nullIndex;
    }
```